keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py

```diff
@@ -0,0 +1,477 @@
+from keras_cv.src.layers.object_detection.box_matcher import BoxMatcher
+from keras_cv.src.layers.object_detection.roi_align import _ROIAligner
+from keras_cv.src.layers.object_detection.roi_generator import ROIGenerator
+from keras_cv.src.layers.object_detection.roi_sampler import _ROISampler
+from keras_cv.src.layers.object_detection.rpn_label_encoder import (
+    _RpnLabelEncoder,
+)
+from keras_cv.src.models.object_detection.__internal__ import unpack_input
+from keras_cv.src.models.object_detection.faster_rcnn import FeaturePyramid
+from keras_cv.src.models.object_detection.faster_rcnn import RCNNHead
+from keras_cv.src.models.object_detection.faster_rcnn import RPNHead
+from keras_cv.src.models.task import Task
+from keras_cv.src.utils.train import get_feature_extractor
+from keras_cv.src import losses
+
+BOX_VARIANCE = [0.1, 0.1, 0.2, 0.2]
+
+
+@keras_cv_export(
+    [
+        "keras_cv.models.FasterRCNN",
+        "keras_cv.models.object_detection.FasterRCNN",
+    ]
+)
+class FasterRCNN(Task):
+    def __init__(
+        self,
+        backbone,
+        num_classes,
+        bounding_box_format,
+        anchor_generator=None,
+        feature_pyramid=None,
+        rcnn_head=None,
+        label_encoder=None,
+        *args,
+        **kwargs,
+    ):
+        # 1. Backbone
+        extractor_levels = ["P2", "P3", "P4", "P5"]
+        extractor_layer_names = [
+            backbone.pyramid_level_inputs[i] for i in extractor_levels
+        ]
+        feature_extractor = get_feature_extractor(
+            backbone, extractor_layer_names, extractor_levels
+        )
+        
+        # 2. Feature Pyramid
+        feature_pyramid = feature_pyramid or FeaturePyramid()
+        
+        # 3. Anchors
+        scales = [2**x for x in [0]]
+        aspect_ratios = [0.5, 1.0, 2.0]
+        anchor_generator = (
+            anchor_generator
+            or FasterRCNN.default_anchor_generator(
+                scales,
+                aspect_ratios, 
+                bounding_box_format,
+                
+            )
+        )
+        
+        # 4. RPN Head
+        num_anchors_per_location = len(scales) * len(aspect_ratios)
+        rpn_head = RPNHead(num_anchors_per_location)
+        # 5. ROI Generator
+        roi_generator = ROIGenerator(
+            bounding_box_format=bounding_box_format,
+            nms_score_threshold_train=float("-inf"),
+            nms_score_threshold_test=float("-inf"),
+            name="roi_generator",
+        )
+        
+        # 6. ROI Pooler
+        roi_pooler = _ROIAligner(bounding_box_format="yxyx", name="roi_pooler")
+        
+        
+        # 7. RCNN Head
+        rcnn_head = rcnn_head or RCNNHead(
+            num_classes,
+            name="rcnn_head"
+        )
+
+        # Begin construction of forward pass
+        image_shape =  feature_extractor.input_shape[1:]
+        if None in image_shape:
+            raise ValueError("Image shape should not have None")
```

Comment:
Problem with this error is that it's so far removed from a user of this class, that this message, while accurate, would make no sense to them.

We should consider when/how such a None could appear in image_shape, and either instead guard against that specific situation or rephrase this error to something the user could understand.

I know this is a bit more work, but it's important I think.

Reply:
Yes, sure makes sense. Generally for us to generate anchors we need Image Shape as must input. I will figure out how can we add more context to this error and make it user friendly.